File: go/web.go
package main

import (
    "fmt"
    "log"
    "net/http"
    "strconv"
    "database/sql"
    _ "github.com/go-sql-driver/mysql"
)

func handler(w http.ResponseWriter, r *http.Request) {
    con, err := sql.Open("mysql", "chen:12345678@/web")
    if err != nil {panic(err)}
    var user_id int
    var username string
    var gender string
    var email string
    userids, ok := r.URL.Query()["gouserid"]
    if !ok || len(userids[0]) < 1 {
        fmt.Println("No input")
        return
    }
    var userid string
    userid = userids[0]
    input, err := strconv.Atoi(userid)
    
    id := con.QueryRow("SELECT * FROM user WHERE user_id=?", input)
    switch err := id.Scan(&user_id, &username, &gender, &email); err {
	case sql.ErrNoRows:
  	    fmt.Println("No rows were returned!")
	case nil:
  	    fmt.Println(username, gender, email)
	default:
  	    panic(err)
    }
	    //fmt.Fprintf(w, "%s!", r.URL.Path[1:])
    defer con.Close();
}

func main() {
    http.HandleFunc("/", handler)
    log.Fatal(http.ListenAndServe(":8080", nil))
}

File: lua/lua/web.lua
local _M = {}


function _M.webservice(id)

    local mysql = require "luasql.mysql"
    local redis = require "redis"
    local client = redis.connect('127.0.0.1', 6379)

    local env = mysql.mysql()
    local con = env:connect('web','chen','12345678')

    cursor,errorString = con:execute([[select user_id,username,gender,email from user]])
    row = cursor:fetch ({}, "a")
    
    while row do
        if (row.user_id == id) then
            if_exist = client:exists(row.user_id)
            ngx.say(if_exist)
            if (if_exist == true) then
                local output = client:get(id)
                ngx.say("Found the key in redis.")
                ngx.say(output)
            else
                local output = row.username .. " " .. row.gender .. " " .. row.email
                client:set(id, output)
                ngx.say("Didn't find the key in redis. Will insert key.")
                ngx.say(output)
            end
	end
	row = cursor:fetch (row, "a")
    end
end

return _M 

